pcsx2/USB/qemu-usb/usb-ohci.cpp

```diff
@@ -0,0 +1,1774 @@
+		{
+			port->ctrl &= ~OHCI_PORT_PES;
+			port->ctrl |= OHCI_PORT_PESC;
+		}
+		dev = port1->dev;
+		if (dev)
+		{
+			dev->port = NULL;
+			/* send the detach message */
+			dev->state = USB_STATE_NOTATTACHED;
+		}
+		port1->dev = NULL;
+	}
+
+	if (old_state != port->ctrl)
+		ohci_set_interrupt(s, OHCI_INTR_RHSC);
+}
+
+/* Attach or detach a device on a root hub port.  */
+static void ohci_attach(USBPort* port1)
+{
+	OHCIState* s = (OHCIState*)port1->opaque;
+	OHCIPort* port = &s->rhport[port1->index];
+	uint32_t old_state = port->ctrl;
+
+	port1->dev->port = port1;
+
+	/* set connect status */
+	port->ctrl |= OHCI_PORT_CCS | OHCI_PORT_CSC;
+
+	/* update speed */
+	if (port->port.dev->speed == USB_SPEED_LOW)
+	{
+		port->ctrl |= OHCI_PORT_LSDA;
+	}
+	else
+	{
+		port->ctrl &= ~OHCI_PORT_LSDA;
+	}
+
+	/* notify of remote-wakeup */
+	if ((s->ctl & OHCI_CTL_HCFS) == OHCI_USB_SUSPEND)
+	{
+		ohci_set_interrupt(s, OHCI_INTR_RD);
+	}
+
+	//trace_usb_ohci_port_attach(port1->index);
+
+	if (old_state != port->ctrl)
+	{
+		ohci_set_interrupt(s, OHCI_INTR_RHSC);
+	}
+}
+
+static void ohci_detach(USBPort* port1)
+{
+	OHCIState* s = (OHCIState*)port1->opaque;
+	OHCIPort* port = &s->rhport[port1->index];
+	uint32_t old_state = port->ctrl;
+
+	if (port1->dev)
+		ohci_async_cancel_device(s, port1->dev);
+
+	/* set connect status */
+	if (port->ctrl & OHCI_PORT_CCS)
+	{
+		port->ctrl &= ~OHCI_PORT_CCS;
+		port->ctrl |= OHCI_PORT_CSC;
+	}
+	/* disable port */
+	if (port->ctrl & OHCI_PORT_PES)
+	{
+		port->ctrl &= ~OHCI_PORT_PES;
+		port->ctrl |= OHCI_PORT_PESC;
+	}
+	//trace_usb_ohci_port_detach(port1->index);
+
+	if (old_state != port->ctrl)
+	{
+		ohci_set_interrupt(s, OHCI_INTR_RHSC);
+	}
+}
+
+static void ohci_wakeup(USBPort* port1)
+{
+	OHCIState* s = (OHCIState*)port1->opaque;
+	OHCIPort* port = (OHCIPort*)&s->rhport[port1->index];
+	uint32_t intr = 0;
+	if (port->ctrl & OHCI_PORT_PSS)
+	{
+		//trace_usb_ohci_port_wakeup(port1->index);
+		port->ctrl |= OHCI_PORT_PSSC;
+		port->ctrl &= ~OHCI_PORT_PSS;
+		intr = OHCI_INTR_RHSC;
+	}
+	/* Note that the controller can be suspended even if this port is not */
+	if ((s->ctl & OHCI_CTL_HCFS) == OHCI_USB_SUSPEND)
+	{
+		//trace_usb_ohci_remote_wakeup(s->name);
+		/* This is the one state transition the controller can do by itself */
+		s->ctl &= ~OHCI_CTL_HCFS;
+		s->ctl |= OHCI_USB_RESUME;
+		/* In suspend mode only ResumeDetected is possible, not RHSC:
+         * see the OHCI spec 5.1.2.3.
+         */
+		intr = OHCI_INTR_RD;
+	}
+	ohci_set_interrupt(s, intr);
+}
+
+static USBDevice* ohci_find_device(OHCIState* ohci, uint8_t addr)
+{
+	USBDevice* dev;
+	int i;
+
+	for (i = 0; i < ohci->num_ports; i++)
+	{
+		if ((ohci->rhport[i].ctrl & OHCI_PORT_PES) == 0)
+		{
+			continue;
+		}
+		dev = usb_find_device(&ohci->rhport[i].port, addr);
+		if (dev != NULL)
+		{
+			return dev;
+		}
+	}
+	return NULL;
+}
+
+//TODO no devices using this yet
+static void ohci_stop_endpoints(OHCIState* ohci)
```

Comment:
You can update this function too, second loop doesn't need unsigned tho but you can move the declarations.